pinot-common/src/main/java/org/apache/pinot/common/utils/helix/TableCache.java

```diff
@@ -18,190 +18,308 @@
+          putSchema(znRecord);
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while adding schema for ZNRecord: {}", znRecord.getId(), e);
+        }
+      }
+    }
+  }
+
+  private void putSchema(ZNRecord znRecord)
+      throws IOException {
+    Schema schema = SchemaUtils.fromZNRecord(znRecord);
+    String rawTableName = schema.getSchemaName();
+    if (_caseInsensitive) {
+      Map<String, String> columnNameMap = new HashMap<>();
+      for (String columnName : schema.getColumnNames()) {
+        columnNameMap.put(columnName.toLowerCase(), columnName);
+      }
+      _schemaInfoMap.put(rawTableName, new SchemaInfo(schema, columnNameMap));
+    } else {
+      _schemaInfoMap.put(rawTableName, new SchemaInfo(schema, null));
+    }
+  }
+
+  private void removeSchema(String path) {
+    _propertyStore.unsubscribeDataChanges(path, _schemaChangeListener);
+    String rawTableName = path.substring(SCHEMA_PATH_PREFIX.length());
+    _schemaInfoMap.remove(rawTableName);
+  }
+
+  private class TableConfigChangeListener implements IZkChildListener, IZkDataListener {
 
     @Override
-    public void handleChildChange(String s, List<String> list)
-        throws Exception {
-      refresh();
+    public synchronized void handleChildChange(String path, List<String> tables) {
+      if (CollectionUtils.isEmpty(tables)) {
+        return;
+      }
+
+      // Only process new added table configs. Changed/removed table configs are handled by other callbacks.
+      List<String> pathsToAdd = new ArrayList<>();
+      for (String tableNameWithType : tables) {
+        if (!_tableConfigMap.containsKey(tableNameWithType)) {
+          pathsToAdd.add(TABLE_CONFIG_PATH_PREFIX + tableNameWithType);
+        }
+      }
+      if (!pathsToAdd.isEmpty()) {
+        addTableConfigs(pathsToAdd);
+      }
     }
 
     @Override
-    public void handleDataChange(String s, Object o)
-        throws Exception {
-      refresh();
+    public synchronized void handleDataChange(String path, Object data) {
+      if (data != null) {
+        ZNRecord znRecord = (ZNRecord) data;
+        try {
+          putTableConfig(znRecord);
+        } catch (Exception e) {
+          LOGGER.error("Caught exception while refreshing table config for ZNRecord: {}", znRecord.getId(), e);
+        }
+      }
     }
 
     @Override
-    public void handleDataDeleted(String s)
-        throws Exception {
-      refresh();
+    public synchronized void handleDataDeleted(String path) {
+      // NOTE: The path here is the absolute ZK path instead of the relative path to the property store.
+      String tableNameWithType = path.substring(path.lastIndexOf('/') + 1);
+      removeTableConfig(TABLE_CONFIG_PATH_PREFIX + tableNameWithType);
     }
   }
 
-  class SchemaChangeListener implements IZkChildListener, IZkDataListener {
-    Map<String, Map<String, String>> _schemaColumnMap = new ConcurrentHashMap<>();
-
-    public synchronized void refresh() {
```

Comment:
this method was good, only thing we had to handle was calling subscribe on each child node. This is a much better way to handle ZK callbacks instead of trying to handle every callback with custom logic

Reply:
We want to avoid redundant ZK accesses. The current way will always read all table configs/schemas on every change. For a cluster with thousands of table, one node change should only requires one ZNRecord fetch (which is already done internally and passed to the callback) instead of thousands.